Expression Matrix Service Charter

ambrosejcarr opened the pull request:
~**October 17**: Last call for oversight reviews~
**October 18**: Approval

ambrosejcarr committed:
upload charter

benedictpaten commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,90 @@
+    * Dependency on Query Service for join queries on metadata
+
+## Out-of-scope
+
```
I appreciate you don't want to create things higher up the tool chain that will be developed by third parties, but to drive development it might help to identify in-scope demos or collaborations that could be used to demonstrate utility. It is important we build something that meets and demonstrates use cases, obviously.

benedictpaten replied:
Perhaps identifying some of these third party clients for testing might be good?

ambrosejcarr replied:
Thanks Benedict! Yes definitely we're super excited about working with developers who want to work with our API so we can make sure we have something useful. 

We're reluctant to name anything specific in the charter though, as it's a pretty sizable expansion of scope to commit to working with particular third party tools who we don't have commensurate commitments from, and we want to remain flexible as the community's needs evolve. 🙂

TimothyTickle replied:
I agree that third-party demos make sense, happy to help source them through the portal group if needed. Let's not hard code demos into the charter, could put something generic outreach for adoption

benedictpaten replied:
No need to commit to specific tools or demos, agree with Tim, something generic at least serves to show intent.

TimothyTickle replied:
Actually, I think this is covered in "Validate design with communities building portals for analysis and visualization" so is ok by me.

ambrosejcarr replied:
@benedictpaten Do you think that the wording as identified by Tim is adequate to serve the intent you're looking for, or would you like me to add another sentence? We definitely _do_ want to make our affirmative intent to collaborate with tool builders clear.

kozbo commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,90 @@
+    * Bundle identifiers
+    * Metadata (*across bundles; potentially provided directly by Query Module*)
+    * Numerical filtering (*on gene expression values, and also QC metrics*)
+    * Compound metadata and numerical filtering
```
I think I know what you mean by each of these points, but will the community? They could easily be expanded with a sentence or two each.

kozbo commented:
Other than a request to elucidate a little more this all looks good to me.

benedictpaten reviewed (approved):
Aside from one comment, lgtm.

ambrosejcarr committed:
elucidated in-scope queries

TimothyTickle commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,93 @@
+
+* Expression Matrix Service interface supports queries based on:
+    * Bundle identifiers
+        * This query enables selection of cells from specific submissions. 
```
Will bundle identifiers enable cell-level selection? or groups of cells?

ambrosejcarr replied:
It's a bundle-level query -- so whatever is in the bundle. I will drop the "cells from" language.

TimothyTickle commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,93 @@
+* Architecture, software design, and scale testing for scalability
+
+* Data representation
+    * Expression matrix data format for transport
```
Does this mean, the format delivered to users? If so, please word more explicitly if possible.

briandoconnor replied:
I'm assuming you'll have a variety of formats available for the user of the API to choose, is that right?

TimothyTickle commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,93 @@
+* Data representation
+    * Expression matrix data format for transport
+    * Expression matrix data format for storage
+    * QC matrix data format for storage and transport
```
Here the format for transport and storage are assumed to be the same format while they are separate for the data matrix. Is this intentional, I am not sure I believe they will be the same format.

TimothyTickle commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,93 @@
+
+# Expression Matrix Service
```
Although we have been loose about the difference between counts and expression matrices, it may be good to tighten that up here. Is this an expression matrix service, a counts matrix service or both? If expression, this assumes UMI collapsing, correction for depth (CPM) and/or transcript length (TPM), I believe we should always deliver counts and optionally deliver other things.

ambrosejcarr replied:
I don't see the "expression" term as overloaded the way you do. I think of an expression matrix as the general term, and count matrices, cpm matrices and tpm matrices are the specific types. 

In any case, I believe we're aligned on delivery of the raw counts, augmented with transcript lengths where appropriate.

TimothyTickle replied:
Maybe just using the term "matrix service" for the charter and not specifying the data type will get around this and make this service to be positioned to be applied for many data modalities long-term.

kbergin replied:
I'm curious, now that the zarr plans include keeping QC metrics calculated as part of the overall matrix outputs, would this need to be generalized to just "matrix service" ?

ambrosejcarr replied:
Interesting idea. The co-delivery of those objects always been my plan. Because the idea and core deliverables remain the same, I don't think there is need for change. 

As a thought experiment, If something not related to expression were to be made into a matrix, I don't think this service would produce it.

kbergin replied:
Works for me!

TimothyTickle commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,93 @@
+    visualization
+    * Provide feedback in the form of architecture sessions and written UX
+    reports to the DCP community about the experience of building layered services using the public API(s). (e.g., *can the expression microservice be built using only public Data Store API?*)
+    * Scientific "guardrails" to warn users when executed queries produce data subsets with questionable scientific utility
```
These are optional given AWG feedback but nice to have as a user experience, this is not required for governance.

TimothyTickle commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,93 @@
+    * Scientific "guardrails" to warn users when executed queries produce data subsets with questionable scientific utility
+
+* External dependencies
+    * Dependencies on Data Store, Query Service, and Secondary Analysis to enable and expose quality control metrics
```
Secondary Analysis also generates data matrices as well as QC metrics, not sure if you want to add that here in the dependencies.

TimothyTickle commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,93 @@
+## Out-of-scope
+
+* Data visualization portal or web application for expression matrices
+* Transformative data manipulation; analysis apps or pipelines using expression data (e.g. data normalization)
```
Does this include CPM and TPM calculations?

ambrosejcarr replied:
We want to make sure to provide useful data to users. Ideally basic normalization approaches calculations would be done by green box, since they are not dataset dependent and can be precomputed. We will work closely with green to identify which types of pre-calculations users and advisors have identified as important.

briandoconnor commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,93 @@
+* Expression Matrix Service interface supports queries based on:
+    * Bundle identifiers
+        * This query enables selection of cells from specific submissions. 
+    * Metadata (*across bundles; potentially provided directly by Query Module*)
```
So just to be clear, the matrix service will index metadata in bundles itself (or use another index like ES or Postgres) to understand metadata queries?  Just asking because on the orange side, we're planning on just handing off a list of bundle IDs/versions but if the matrix service "understands" the HCA metadata then more interesting query handoff may be possible.  On the orange side we'd need to understand more about the query abilities for the matrix service but for now I think the handoff of bundle UUIDs and versions is sufficient.

briandoconnor commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,93 @@
+
+* User Experience
+    * Conduct UX research and design; drive query and API design directly from use cases
+    * Validate design with communities building portals for analysis and
```
It would be useful for portal developers to get information on how to leverage the matrix service as well as the faceted data browser in Orange box for their own portals.  We should really explore this more, I'm thinking preparing materials for training together (and make sure the Orange box charter mentioned data browser reuse by portal developers too)  @theathorn FYI

briandoconnor commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,93 @@
+    visualization
+    * Provide feedback in the form of architecture sessions and written UX
+    reports to the DCP community about the experience of building layered services using the public API(s). (e.g., *can the expression microservice be built using only public Data Store API?*)
+    * Scientific "guardrails" to warn users when executed queries produce data subsets with questionable scientific utility
```
I can understand adding this to the charter but it's pretty ambiguous so what does this mean exactly?

ambrosejcarr replied:
Added sentence to clarify.

briandoconnor commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,93 @@
+    * QC matrix data format for storage and transport
+
+
+* Extends HCA DCP Command Line Interface (CLI) client to support the Expression Matrix Service
```
Mention here working on a handoff mechanism so queries in Orange data browser can be handed to matrix service.  I think this will prove to be a very popular way to find data in the matrix service and generate a result, perhaps more so than queries written  with the CLI.

briandoconnor commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,93 @@
+
+## Out-of-scope
+
+* Data visualization portal or web application for expression matrices
```
OK, but I think what is in scope is coordinating query handoff from Orange to the matrix service.  We've already talked with Marcus about how this would work.  But I don't see it in the charter.  Can this interaction from Orange to the Matrix service be specifically called out as in scope?  I added a note above around the CLI bulletpoint for in scope items.

brianraymor replied:
@briandoconnor - Based on the updated plan, this comment is no longer relevant and should not block approval. See my response in your similar note above.

briandoconnor reviewed (changes requested):
A few small questions.  Primary missing item I feel is the link to orange box.  This charter specifically says expanding the HCA CLI to support the matrix service.  But I think it's important to mention supporting the "handoff mechanism" from the Orange Box Data Browser as well.

@brianraymor responds:

@briandoconnor - Based on the updated [plan](https://app.zenhub.com/workspace/o/humancellatlas/data-browser/issues/9#issuecomment-416407157), this comment is no longer relevant and should not block approval?

mckinsel commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,93 @@
+
+### Project Lead
+
+[Ambrose Carr](mailto:acarr@chanzukerberg.com)
```
Misspelled zuckerberg

ambrosejcarr committed:
respond to comments

lauraclarke commented:
@ambrosejcarr @brianraymor @mckinsel I am concerned with the out of scope sentence

"Standards engagement for APIs or data storage formats with the GA4GH Large-Scale Genomics work stream"

I am sure you didn't mean it this way but it reads like you will neither engage with or observe GA4GH standards for expression matrices but given our role as a GA4GH driver project that doesn't seem like a stance we can take https://www.ga4gh.org/howwework/driver-projects.html

Can you be more precise about what you intended here?

Brian responds:

@lauraclarke - To be more precise, standards are out-of-scope until we have more experience with "running code" as a source of substantive guidance. Clarifying:

_Standards engagement for APIs or data storage formats with the GA4GH Large-Scale Genomics work stream **prior to basic completion of matrix service implementation and integration of user feedback**_

brianraymor committed:
Clarified GA4GH out-of-scope

* Clarified GA4GH out-of-scope to address Laura's feedback
* Editorial: Removed external dependency which had already been struck

brianraymor committed:
Updates to section headings

The charter template was updated - #18

brianraymor committed:
Clarify "transport" formats

brianraymor committed:
Moved compound filters to out-of-scope

brianraymor committed:
Refine milestones and deliverables

kbergin commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,106 @@
+
+## In-scope
+
+* Expression Matrix Service interface supports queries based on:
```
Does any of this need to be updated after the discussion on pipelines call today?

brianraymor replied:
No updates are required.

kbergin commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,106 @@
+
+## Milestones and Deliverables
+
+**Nov 2018**: *Deploy a minimum viable product (MVP) feature set as part of HCA DCP Beta.*
```
And maybe this will need to be updated per discussion today?

brianraymor replied:
No updates are required.

brianraymor committed:
Updated milestones for delivery formats

Community standards for matrix formats (such as loom or parquet) are being delivered in both v0 and v1. Original plan was to deliver all in v0.

lauraclarke commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,107 @@
+    * Scientific "guardrails" to warn users when executed queries produce data subsets with questionable scientific utility
+        * For example, if a user queries for the expression of only a single gene across all lung cells, this query would trigger a warning that the returned data cannot be effectively normalized
+
+* External dependencies
```
Is there an explicit dependency on the metadata schema team? We call out to gathering requirements from and supporting data consumers to extend the schemas to meet their needs. I would assume you are a specific type data consumer in this context and will be able to provide us with exemplar queries that people want to do.

brianraymor replied:
I reviewed with Ambrose and Marcus. We don't expect there to be an explicit dependency. Any related requests on Metadata would be from Secondary Analysis.

lauraclarke commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,107 @@
+* Compound metadata and numerical filtering (*combinations of the In-scope filters*)
+* Data visualization portal or web application for expression matrices
+* Transformative data manipulation; analysis apps or pipelines using expression data (e.g. data normalization)
+* Standards engagement for APIs or data storage formats with the GA4GH Large Scale Genomics work stream prior to basic completion of matrix service implementation and integration of user feedback
```
Just a quick check on this statement, I assume the caveat "prior to basic completion of matrix service implementation and integration of user feedback" means when the service is up and running and people are actively using it you will recharter to move standards engagement into scope?

brianraymor replied:
We would re-charter and move standards engagement into scope if it aligned with the strategic objectives of the HCA as expressed by guidance from the GG.

jodihir commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,107 @@
+## Description
+
+The expression matrix service enables researchers to easily access and analyze HCA expression matrix data.
+
```
Can you add some description about what the service specifically does; ie. something about how it takes requests from users for specific data and delivers a matrix of expression values.. in some format or formats. The sentence here just says why it's useful.

brianraymor replied:
@jodihir - This is covered in the **Objectives** that follow, but I agree the **Description** is a bit of koan. I could ~steal~ borrow from a document that @mckinsel has been writing:

_The Human Cell Atlas (HCA) Data Coordination Platform (DCP) ingests, processes, and stores single-cell data, making that data easily available to a wide audience of researchers. The single-cell data type of most interest to researchers is the expression matrix, the cell-by-gene expression values that are the starting point for many downstream analyses. To support this data type, the Expression Matrix Service defines a public API to return expression matrices in response to queries._

Thoughts?

jodihir replied:
sounds good

brianraymor committed:
Added mailing list to Communication section

brianraymor committed:
Updated Description based on reviews

brianraymor committed:
Editorial updates to Slack Channels

Adopt best practices for slack channel format

brianraymor committed:
Added descriptions to the Github repositories

lauraclarke commented on charters/ExpressionMatrixService/charter.md:
```diff
@@ -0,0 +1,110 @@
+
+* Serve as a test bed for a re-usable metadata query service, created by
+the *Query Service* component.
+
```
I appreciate I have missed the boat on community review so feel free to ignore this comment but this feels like an odd objective to have in the charter.

I totally get that the matrix service will possibly be the first user of the query service but to explicitly charter that it will be a test bed seems odd as this seems like a much more tightly timeboxed task than most other things mentioned in this charter.
